### PR TITLE
redirect error report to slack channel

### LINKED ISSUE #1449: Send monitoring reports which contain system error to Slack channel instead of email
Currently, monitoring reports are sent through email. We need to redirect the reports to Slack when there is a system error.

### PR DESCRIPTION
**Pull request by candicegjing**
Fixed #1449 
Still need to decide whose token to use.

### CODE REVIEW

File: monitor.py [hunk 1]
```diff
@@ -1,16 +1,14 @@
-#!/usr/bin/env python
+#!/usr/bin/python3
```
Comment (percyliang) [inline monitor.py]: I think using `/usr/bin/env` is better

File: monitor.py [hunk 2]
```diff
@@ -104,6 +105,32 @@ def send_email(subject, message):
     s.quit()
 
 
+def send_slack_notification(subject, message):
```
Comment (percyliang) [inline monitor.py]: Should factor this out into a separate `slack.py` file that can be used by other parts of the codebase.

File: monitor.py [hunk 3]
```diff
@@ -140,9 +167,9 @@ def error_logs(error_type, s):
     last_t = last_sent[error_type]
     t = time.time()
 
-    # Send email only every 4 hours
+    # Send Slack notification only every 4 hours
     if t > last_t + 60 * 60 * 4:
-        send_email('%s [%d times]' % (error_type, n), s.split('\n'))
```
Comment (percyliang) [inline monitor.py]: I don't think we want to get rid of email.  There are some installations where people don't use slack

**Review by percyliang — CHANGES_REQUESTED** [review]
There are other parts of the codebase that send email (e.g., when there are exceptions thrown), which should have slack added.

**Comment by epicfaace** [comment]
Looking into this and maybe another way to do this is to _keep_ the e-mail integration. Slack can give you a special email address, and if you send emails to that address, the messages will show up in a particular channel (https://codalab.slack.com/apps/A0F81496D-email)

Not sure which way is better, though. It might be easier for deployment if we don't have to manage more variables such as the slack API key, but on the other hand, if we are going to have more deep slack integration in the future, it would be good to continue integrating directly via Slack's API.

(it's worth noting that the Email add-on is available only to workspaces with a "Slack Standard Plan" or above, which won't be a problem for us, but might be a problem if someone else with a free slack workspace wants to integrate with it)

**Comment by percyliang** [comment]
Oh, I think if we can just do it via email, maybe we should just do that, since it minimizes the number of dependencies we have?

**Comment by candicegjing** [comment]
Yes, I agree. If we can do it through email, we should use it. @percyliang Then for this ticket, should I just close this PR and set up the email configuration?

**Comment by percyliang** [comment]
Yes, maybe convert this ticket into just some documentation on how to setup
Slack via email.

**Comment by candicegjing** [comment]
Screen Shot 2019-10-11 at 1 21 56 PM

**Review by epicfaace — COMMENTED** [review]
Will we need to modify https://github.com/codalab/codalab-worksheets/blob/master/monitor.py too?

**Review by percyliang — CHANGES_REQUESTED** [review]
I don't think we need a code change here...just change our admin_email to point to slack.

File: docker/compose_files/docker-compose.yml [hunk 4]
```diff
@@ -133,6 +133,22 @@ services:
       - service
       - worker
 
+  worker-shared-file-system:
```
Comment (percyliang) [inline docker/compose_files/docker-compose.yml]: Why does this show up in this PR?

File: docs/Server-Setup.md [hunk 5]
```diff
@@ -249,4 +249,21 @@ The issue is documented in detail [here](https://github.com/benoitc/gunicorn/iss
 is to unselect `/tmp` in Docker's preferences under file sharing.
 
 The image below shows where the file sharing pane is.
-<img src="images/docker-preferences.png" />
\ No newline at end of file
+<img src="images/docker-preferences.png" />
+
+## Sending Slack notification in monitoring service
+
+If you need to send Slack notification from monitoring service, there are two ways to configure your system.
```
Comment (percyliang) [inline docs/Server-Setup.md]: "Slack notifications from the `monitor.py` service"

File: docs/Server-Setup.md [hunk 6]
```diff
@@ -249,4 +249,21 @@ The issue is documented in detail [here](https://github.com/benoitc/gunicorn/iss
...
 
 The image below shows where the file sharing pane is.
-<img src="images/docker-preferences.png" />
\ No newline at end of file
+<img src="images/docker-preferences.png" />
+
+## Sending Slack notification in monitoring service
+
+If you need to send Slack notification from monitoring service, there are two ways to configure your system. 
+1. Using Slack rest API
```
Comment (percyliang) [inline docs/Server-Setup.md]: We don't support this, right?  Delete?

File: docs/Server-Setup.md [hunk 7]
```diff
@@ -249,4 +249,21 @@ The issue is documented in detail [here](https://github.com/benoitc/gunicorn/iss
...
+     * System prerequisite: install slackclient package on the platform where hosts monitoring service
+                `pip3 install slackclient`  
+     * Generate Slack API token via [create-api-token](https://slack.com/help/articles/215770388-create-and-regenerate-api-tokens)
+     * Figure out which channel you want to send notification to: SlACK_CHANNEL_ID
+     * Integrate your Slack messaging function to monitoring service: [Slack API basic_usage](https://slack.dev/python-slackclient/basic_usage.html)
+     * Start your monitoring service
+     * Note that this method doesn't have any restriction on your Slack plans.
+1. Using Slack Email App
```
Comment (percyliang) [inline docs/Server-Setup.md]: Give more context: say "System notifications (e.g., monitor.py) are sent to `$ADMIN_EMAIL`.  If you're using Slack, then you can set $ADMIN_EMAIL to a special email address which will show up in a designated slack channel.  See X for more details.  Note that this works only with the standard plan."

**Review by percyliang — CHANGES_REQUESTED** [review]
I think this should not be labeled as ADMIN versus SLACK but rather as MONITOR (automatic messages that we don't reply to - happen to go to slack) and HELP (messages from users - happen to go to gmail).

**Review by percyliang — COMMENTED** [review]
The prod/dev distinction should not appear in the codalab-worksheets repo.  It should be a configuration in our deployment repo.

File: codalab_service.py [hunk 8]
```diff
@@ -215,7 +215,8 @@ def has_callable_default(self):
         default=100,
     ),
     ### Email
-    CodalabArg(name='admin_email', help='Email to send admin notifications to (e.g., monitoring)'),
+    CodalabArg(name='monitor_email', help='Monitoring email to send notifications to'),
+    CodalabArg(name='help_email', help='Help email to send user questions to'),
```
Comment (percyliang) [inline codalab_service.py]: Maybe call this `admin_email` and `support_email` to use consistent terminology with what's used in the `config.json` file.

**Comment by candicegjing** [comment]
Closed via #1449

**Pull request merged**